### PR TITLE
Writing policies

### PR DESCRIPTION
**Pull request by ereslibre**
Recreated https://github.com/kubewarden/docs/pull/14 after it got closed by the master -> main rename.

Revisit the "writing policies" page
    
* Explain what kubewarden policies are
* Provide a low level overview of the communication protocol
  used to communicate with policies. This is useful when new languages
  are going to be added, new SDKs are going to be written,...

This PR is built on top of https://github.com/kubewarden/docs/pull/15

This PR addresses this issue: #1

### CODE REVIEW

File: src/writing-policies/index.md [hunk 1]
```diff
@@ -1,12 +1,65 @@
...
+# What is a Kubewarden policy
 
-Kubewarden policies can be written using any kind of language capable of building
-[Wasm](https://webassembly.org/) binaries and that supports [waPC](https://github.com/wapc) guest SDK.
+In this section we will explain what Kubewarden policies are by using some traditional computing
+analogies.
 
-> **Note well:** currently Kubewarden supports only Validating Admission Webhooks,
-> Mutating ones are not yet implemented.
+Kubewarden can be seen as a regular program that does one job: it receives
```
Comment (ereslibre) [inline src/writing-policies/index.md:6]:
```suggestion
A Kubewarden policy can be seen as a regular program that does one job: it receives
```

File: src/writing-policies/index.md [hunk 2]
```diff
@@ -1,12 +1,65 @@
...
+input data, performs some computation against that and it finally returns a response.
 
-## Writing a validation policy
+The input data are Kubernetes admission requests and the result of the computation
+is a validation response, something that tells to Kubebernetes whether to accept, reject or
```
Comment (ereslibre) [inline src/writing-policies/index.md:10]:
```suggestion
is a validation response, something that tells to Kubernetes whether to accept, reject or
```

File: src/writing-policies/index.md [hunk 3]
```diff
@@ -1,12 +1,65 @@
...
+accepted a mutated version of the original input data.
```
Comment (ereslibre) [inline src/writing-policies/index.md:11]:
```suggestion
mutate the original input data.
```

File: src/writing-policies/index.md [hunk 4]
```diff
@@ -1,12 +1,65 @@
...
+
+All these operations are performed by a component of Kubewarden that is called
+[policy-server](https://github.com/kubewarden/policy-server).
+
+Policy server doesn't bundle any data processing capability. All these capabilities are
```
Comment (ereslibre) [inline src/writing-policies/index.md:16]:
```suggestion
The policy server doesn't bundle any data processing capability. All these capabilities are
```

File: src/writing-policies/index.md [hunk 5]
```diff
@@ -1,12 +1,65 @@
...
+As a consequence, a Kubewarden policy can be seen as a [traditional plug-in](https://en.wikipedia.org/wiki/Plug-in_%28computing%29)
+of the "policy server" program.
+
+To recap:
+
+  * Kubewarden policies are plug-ins that expose a set of well-defined
+    functionalities (validate a Kubernetes request object, validate policy settings
+    provided by the user,...) using a well-defined API
+  * Policy server is the "main" program that loads the plug-ins
+    (aka policies) and leverages their exposed functionalities to validate
```
Comment (ereslibre) [inline src/writing-policies/index.md:28]:
```suggestion
    (aka policies) and leverages their exposed functionalities to validate or mutate
```

File: src/writing-policies/index.md [hunk 6]
```diff
@@ -1,12 +1,65 @@
...
+    Kubernetes requests
+
+Writing a Kubewarden policies consists of writing the validation business logic
```
Comment (ereslibre) [inline src/writing-policies/index.md:31]:
```suggestion
Writing Kubewarden policies consists of writing the validation business logic
```

File: src/writing-policies/index.md [hunk 7]
```diff
@@ -1,12 +1,65 @@
...
+provides a nice overview of the WebAssembly landscape.
+
+Currently WebAssembly doesn't have an official way to share complex data types
+between the host and a WebAssembly guest. To overcome this limitation
+Kubewarden policies leverage the [waPC](https://github.com/wapc) project, which
+provides a bi-directional communication channel.
+
+Because of that your programming language of choice must provide a waPC guest SDK.
+If that's not the case, feel free to reach out. We can help you overcome this
+limit.
```
Comment (ereslibre) [inline src/writing-policies/index.md:51]:
```suggestion
limitation.
```

File: src/writing-policies/index.md [hunk 8]
```diff
@@ -24,31 +77,38 @@ a function named `validate`, exposed through the waPC guest SDK, so
...
 
 The `validate` function receives a `ValidationRequest` object serialized as JSON and
-returns a `ValidationResponse` object serialized as JSON. 
+returns a `ValidationResponse` object serialized as JSON.
 
 ### The `ValidationRequest` object
 
-The `ValidationRequest` is a simple JSON object that looks like that:
+The `ValidationRequest` is a simple JSON object that is received by the
+`validate` function. It looks like that:
```
Comment (ereslibre) [inline src/writing-policies/index.md:85]:
```suggestion
`validate` function. It looks like this:
```
?

File: src/writing-policies/index.md [hunk 9]
```diff
@@ -24,31 +77,38 @@ a function named `validate`, exposed through the waPC guest SDK, so
...
+specific settings.
+
 ### The `ValidationResponse` object
 
-The `ValidationResponse` object is a simple JSON object like the
-following:
+The `validate` function returns the outcome of its validation using a `ValidationResponse`
+object.
+
+The `ValidationResponse` is structured in this way:
```
Comment (ereslibre) [inline src/writing-policies/index.md:104]:
```suggestion
The `ValidationResponse` is structured in the following way:
```

File: src/writing-policies/index.md [hunk 10]
```diff
@@ -65,21 +125,35 @@ accepted, the Kubernetes API server will forward this information as
...
-The list of languages that can produce Wasm modules is continuously evolving.
-[This page](https://github.com/appcypher/awesome-wasm-langs) provides a
-nice overview.
+If the user provided settings are `valid`, the contents of `message` are ignored. Otherwise
+the contents of `message` are shown to the user.
 
-It's possible to check what languages support waPC guest by searching
-on the [official github
-group](https://github.com/wapc?q=guest&type=&language=).
+Kubewarden's policy server validates of all the policy settings provided by users at start time.
```
Comment (ereslibre) [inline src/writing-policies/index.md:158]:
```suggestion
Kubewarden's policy server validates all the policy settings provided by users at start time.
```

**Review by ereslibre — COMMENTED** [review]
Thank you @flavio, some comments.

Also, with the branch rename I think you might want to rebase on top of main, since when fixing it I have the impression main has already some changes shown in this PR... check commit https://github.com/kubewarden/docs/commit/824b81103fe343cd44a0090d2d667863b4d13ae0

**Comment by flavio** [comment]
@ereslibre I've included all your suggestions (thanks!), squashed all of them into a single commit (with you as co-author) and rebased. I've dropped the commit about the "testing policies" section, there's another PR for that.

**Pull request merged**